docs/man/es/man1/hy_gt_vfd.1
'\" t
.\"     Title: hy_gt_vfd
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 08/11/2020
.\"    Manual: LinuxCNC Documentation
.\"    Source: LinuxCNC
.\"  Language: English
.\"
.TH "HY_GT_VFD" "1" "08/11/2020" "LinuxCNC" "LinuxCNC Documentation"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
hy_gt_vfd \- HAL userspace component for Huanyang GT\-series VFDs
.SH "SYNOPSIS"
.sp
\fBhy_gt_vfd\fR [\fIOPTIONS\fR]
.SH "DESCRIPTION"
.sp
The hy_gt_vfd component interfaces a Huanyang GT\-series VFD to the LinuxCNC HAL\&. The VFD is connected via RS\-485 serial to the LinuxCNC computer\&.
.SH "HARDWARE SETUP"
.sp
At least some Huanyang GT VFDs must be physically modified to enable Modbus communication\&.
.sp
The circuit board location marked "SW1" is identified in the manual as "Switch of terminal resistor for RS485 communication"\&. On the only VFD I have experience with, the circuit board contained no switch at that location, instead holding a pair of crossed jumper wires (top\-left pad connected to bottom\-right pad, top\-right to bottom\-left)\&. In this configuration, no Modbus communication is possible\&. We had to desolder the two crossed jumper wires and re\-solder them parallel to each other (top\-left to bottom\-left, top\-right to bottom\-right)\&.
.SH "FIRMWARE SETUP"
.sp
The Huanyang GT VFD must be configure via the faceplate to talk Modbus with LinuxCNC\&. Consult the Operation section of the Huanyang GT\-series Inverter Manual for details\&. Set the following parameters:
.PP
P0\&.01 = 2
.RS 4
Set Run Command Source to Modbus serial port\&.
.RE
.PP
P0\&.03
.RS 4
Set Maximum Frequency to the maximum frequency you want the VFD to output, in Hz\&.
.RE
.PP
P0\&.04
.RS 4
Set Upper Frequence Limit to the maximum frequency you want the VFD to output, in Hz\&. This should be the same as the value in P0\&.03\&.
.RE
.PP
P0\&.05
.RS 4
Set Lower Frequency Limit to the minimum frequency you want the VFD to output, in Hz\&.
.RE
.PP
P0\&.07 = 7
.RS 4
Set Frequency A Command Source to Modbus serial port\&.
.RE
.PP
P2\&.01 = ???
.RS 4
Set Motor Rated Power to the motor\(cqs power rating in kW\&.
.RE
.PP
P2\&.02 = ???
.RS 4
Set Motor Rated Frequency to the motor\(cqs max frequency in Hz\&.
.RE
.PP
P2\&.03 = ???
.RS 4
Set Motor Rated Speed to the motor\(cqs speed in RPM at its rated maximum frequence\&.
.RE
.PP
P2\&.04 = ???
.RS 4
Set Motor Rated Voltage to the motor\(cqs maximum voltage, in Volts\&.
.RE
.PP
P2\&.05 = ???
.RS 4
Set Motor Rated Current to the motor\(cqs maximum current, in Amps\&.
.RE
.PP
PC\&.00 = 1
.RS 4
Set Local Address to 1\&. This matches the default in the hy_gt_vfd driver, change this if your setup has special needs\&.
.RE
.PP
PC\&.01 = 5
.RS 4
Set Baud Rate Selection to 5 (38400 bps)\&. This matches the default in the hy_gt_vfd driver, change this if your setup has special needs\&.
.sp
.if n \{\
.RS 4
.\}
.nf
0 = 1200
1 = 2400
2 = 4800
3 = 9600
4 = 19200
5 = 38400
.fi
.if n \{\
.RE
.\}
.RE
.PP
PC\&.02 = 0
.RS 4
Set Data Format (8n1 RTU)\&. This matches the default in the hy_gt_vfd driver, change this if your setup has special needs\&.
.RE
.PP
PC\&.03 = 1
.RS 4
Set Communication Delay Time to 1 ms\&. This is expected by the hy_gt_vfd driver\&.
.RE
.SH "OPTIONS"
.PP
\fB\-b\fR, \fB\-\-bits\fR \fIN\fR
.RS 4
(default 8) For Modbus communication\&. Set number of data bits to
\fIN\fR\&.
\fIN\fR
must be between 5 and 8 inclusive\&.
.RE
.sp
\fB\-p\fR, \fB\-\-parity\fR [Even,Odd,None]
.sp
.if n \{\
.RS 4
.\}
.nf
(default None) For Modbus communication\&.  Set serial parity to Even,
Odd, or None\&.
.fi
.if n \{\
.RE
.\}
.PP
\fB\-r\fR, \fB\-\-rate\fR \fIN\fR
.RS 4
(default 38400) For Modbus communication\&. Set baud rate to
\fIN\fR\&. It is an error if the rate is not one of the following: 1200, 2400, 4800, 9600, 19200, 38400
.RE
.PP
\fB\-s\fR, \fB\-\-stopbits\fR [1,2]
.RS 4
(default 1) For Modbus communication\&. Set serial stop bits to 1 or 2\&.
.RE
.PP
\fB\-t\fR, \fB\-\-target\fR \fIN\fR
.RS 4
(default 1) For Modbus communication\&. Set Modbus target (slave) number\&. This must match the device number you set on the Huanyang GT VFD\&.
.RE
.PP
\fB\-d\fR, \fB\-\-device\fR \fIPATH\fR
.RS 4
(default /dev/ttyS0) For Modbus communication\&. Set the name of the serial device node to use\&.
.RE
.PP
\fB\-v\fR, \fB\-\-verbose\fR
.RS 4
Turn on verbose mode\&.
.RE
.PP
\fB\-S\fR, \fB\-\-motor\-max\-speed\fR \fIRPM\fR
.RS 4
The motor\(cqs max speed in RPM\&. This must match the motor speed value configured in VFD register P2\&.03\&.
.RE
.PP
\fB\-F\fR, \fB\-\-max\-frequency\fR \fIHZ\fR
.RS 4
This is the maximum output frequency of the VFD in Hz\&. It should correspond to the motor\(cqs rated max frequency, and to the maximum and upper limit output frequency configured in VFD register P0\&.03 and P0\&.04\&.
.RE
.PP
\fB\-f\fR, \fB\-\-min\-frequency\fR \fIHZ\fR
.RS 4
This is the minimum output frequency of the VFD in Hz\&. It should correspond to the minimum output frequency configured in VFD register P0\&.05\&.
.RE
.SH "PINS"
.PP
\fBhy_gt_vfd\&.period\fR (float, in)
.RS 4
The period for the driver\(cqs update cycle, in seconds\&. This is how frequently the driver will wake up, check its HAL pins, and communicate with the VFD\&. Must be between 0\&.001 and 2\&.000 seconds\&. Default: 0\&.1 seconds\&.
.RE
.PP
\fBhy_gt_vfd\&.speed\-cmd\fR (float, in)
.RS 4
The requested motor speed, in RPM\&.
.RE
.PP
\fBhy_gt_vfd\&.speed\-fb\fR (float, out)
.RS 4
The motor\(cqs current speed, in RPM, reported by the VFD\&.
.RE
.PP
\fBhy_gt_vfd\&.at\-speed\fR (bit, out)
.RS 4
True when the drive is on and at the commanded speed (within 2%), False otherwise\&.
.RE
.PP
\fBhy_gt_vfd\&.freq\-cmd\fR (float, out)
.RS 4
The requested output frequency, in Hz\&. This is set from the \&.speed\-cmd value, and is just shown for debugging purposes\&.
.RE
.PP
\fBhy_gt_vfd\&.freq\-fb\fR (float, out)
.RS 4
The current output frequency of the VFD, in Hz\&. This is reported from the VFD to the driver\&.
.RE
.PP
\fBhy_gt_vfd\&.spindle\-on\fR (bit, in)
.RS 4
Set this pin True to command the spindle on, at the speed requested on the \&.speed\-cmd pin\&. Set this pin False to command the spindle off\&.
.RE
.PP
\fBhy_gt_vfd\&.output\-voltage\fR (float, out)
.RS 4
The voltage that the VFD is current providing to the motor, in Volts\&.
.RE
.PP
\fBhy_gt_vfd\&.output\-current\fR (float, out)
.RS 4
The current that the motor is currently drawing from the VFD, in Amperes\&.
.RE
.PP
\fBhy_gt_vfd\&.output\-power\fR (float, out)
.RS 4
The power that the motor is currently drawing from the VFD, in Watts\&.
.RE
.PP
\fBhy_gt_vfd\&.dc\-bus\-volts\fR (float, out)
.RS 4
The current voltage of the VFD\(cqs internal DC power supply, in Volts\&.
.RE
.PP
\fBhy_gt_vfd\&.modbus\-errors\fR (u32, out)
.RS 4
A count of the number of modbus communication errors between the driver and the VFD\&. The driver is resilient against communication errors, but a large or growing number here indicates a problem that should be investigated\&.
.RE
.PP
\fBhy_gt_vfd\&.input\-terminal\fR (float, out)
.RS 4
The VFD\(cqs input terminal register\&.
.RE
.PP
\fBhy_gt_vfd\&.output\-terminal\fR (float, out)
.RS 4
The VFD\(cqs output terminal register\&.
.RE
.PP
\fBhy_gt_vfd\&.AI1\fR (float, out)
.RS 4
The VFD\(cqs AI1 register\&.
.RE
.PP
\fBhy_gt_vfd\&.AI2\fR (float, out)
.RS 4
The VFD\(cqs AI2 register\&.
.RE
.PP
\fBhy_gt_vfd\&.HDI\-frequency\fR (float, out)
.RS 4
The VFD\(cqs HDI\-frequency register\&.
.RE
.PP
\fBhy_gt_vfd\&.external\-counter\fR (float, out)
.RS 4
The VFD\(cqs external counter register\&.
.RE
.PP
\fBhy_gt_vfd\&.fault\-info\fR (float, out)
.RS 4
The VFD\(cqs fault info register\&.
.RE
.SH "ISSUES"
.sp
The VFD produces the output frequency that it sends to the motor by adding a manually specified offset to the frequency command it gets over modbus\&.
.sp

docs/NEWS
check debian/changelog for recent news

# some of EMC2 History & News
#############################

2009.06.13 - mshaver added emc2/configs/smithy directory and related files
             for Smithy Company's line of cnc machines.

2006.11.17 - removed HAL types s8, u8, s16, and u16.  HAL now uses float
	     for analog values, bit for booleans, and s32 or u32 for
	     integer values.  The only users who will be affected are
	     those with Universal Stepper Controller cards.  See item 9 at
	     http://wiki.linuxcnc.org/cgi-bin/emcinfo.pl?UpdatingConfigurationsForDevelopmentVersions

2006.11.15 - merged classicladder 0.7.100

2006.11.04 - added a feed-hold HAL pin.  Added M codes to enable/disable
	     feed override, spindle override, feed hold, and adaptive feed
	     individually.  Feed override is M50, spindle override is M51,
	     adaptive feed is M52, and feed hold is M53.  G50/51 are no
	     longer used to enable/disable adaptive feed, use M52 instead.

2006.10.24 - added support for block delete as defined by Section 2.2.2
             in the RS274NGC_3 specs.

2006.10.24 - added support for optional program stop as defined by Section 
             2.2.3 in the RS274NGC_3 specs.

2006.10.14 - manpages have been created for most rtapi and hal calls

2006.10.01 - added a userspace simulator which includes rtapi threads.
	     Use "--enable-simulator" to enable this mode, which cannot
	     be used to control real hardware

2006.09.27 - added support to optionally save joint positions between
             runs ([TRAJ] POSITION_FILE = position.txt)

2006.09.26 - added support to specify UNITS in literal in the ini. 
             (e.g. [TRAJ] LINEAR_UNITS = mm )

2006.09.09 - fixed bug #1384883, optional stop was not optional.

2006.08.09 - AXIS is now a part of EMC2, not a separate source download.

2006.08.07 - Added a spindle-speed override. This allows to override the spindle
	     speed during a program run. The MIN_SPINDLE_OVERRIDE and 
	     MAX_SPINDLE_OVERRIDE in the ini assure that speed doesn't get 
	     reduced or increased too much. Both values are in %.

2006.08.06 - Documentation can be built from source with the 'make docs' target

2006.06.11 - added realtime support for jogwheels.  Three HAL pins per axis:
	     one accepts raw counts from the wheel, one sets the distance moved
	     per count, and one enables that axis to move.

2006.05.17 - added 'adaptive feed', a HAL pin that can be used to scale the
	     feedrate in real time.  Simular to feedrate override, but can
	     only vary from 0.0 to 1.0, and reacts much faster.  Intended 
	     for use with EDM and other applications that require closed 
	     loop control of feedrate.

2006.03.18 - A new pin "count" exists for freqgen.  It is the same as the
             internal symbol "count" shown in the hal introduction.pdf,
             or the result of up OR down if there was an "OR" component.

2006.03.16 - Modified the motion control module to behave more like other
             HAL modules - it no longer adds its functions to threads
             automatically.  All the sample configs have been changed to
             explicitly add the functions, users who have custom configs
             will need to do the same.

2006.03.12 - If a copy of axis 1.3a0 or newer is detected inside src/
             it will be automatically built and installed for
             --enable-run-in-place use

2006.03.10 - minor bugfix in tkemc for jogging

2006.03.10 - added genhexkins, with a minitetra sample config.
             This proves EMC2 works just as good with hexapods as EMC1 did.

2006.03.10 - added VTI support & sample configs

2006.03.07 - lots of fixes to the TP (most were wrong in the old one as well)

2006.03.05 - new Trajectory Planner

2006.02.20 - added a manpage for the runscript.

2006.02.12 - Fixed a bug that was causing the RT code to run very slowly,
             forcing users to set BASE_PERIOD very long.

2006.02.10 - The build process has been replaced, no more recursive $(MAKE)

2006.01.28 - fixed bug #1205237, mode switching sometimes hung

2006.01.08 - M62/M63 digital outputs synchronous to motion
             M64/M65 digital outputs that get set right away

2006.01.06 - work on halconfig.tcl, it is used to display/configure HAL data

src/hal/user_comps/hy_gt_vfd.c
/*
    hy_gt_vfd.c

    This is a userspace program that interfaces Huanyang GT-series VFDs
    to the LinuxCNC HAL.

    Copyright (C) 2017 Sebastian Kuzminsky

    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU Lesser General Public
    License as published by the Free Software Foundation, version 2.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.

    You should have received a copy of the GNU Lesser General Public
    License along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301-1307 USA.
*/

#include <errno.h>
#include <getopt.h>
#include <math.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <modbus.h>

#include "hal.h"
#include "rtapi.h"


// If a modbus transaction fails, retry this many times before giving up.
#define NUM_MODBUS_RETRIES 5


typedef struct {
    hal_float_t *period;

    hal_float_t *speed_cmd;
    hal_float_t *freq_cmd;
    hal_bit_t *at_speed;

    hal_bit_t	*spindle_on;

    hal_u32_t *modbus_errors;
} haldata_t;

haldata_t *haldata;

int hal_comp_id;


typedef struct {
    int address;  // 0-0xffff for a register, -1 for end-of-list
    const char *name;

    // Multiply the uint16 value read by this multiplier to get the human-
    // readable value.
    float multiplier;

    hal_float_t *hal_pin;
} modbus_register_t;

int num_modbus_registers = 0;
modbus_register_t *modbus_register;


static int done;
char *modname = "hy_gt_vfd";

float motor_max_speed = 0.0;
float max_freq = 0.0;
float min_freq = 0.0;

int baud;

static struct option long_options[] = {
    {"device", 1, 0, 'd'},
    {"rate", 1, 0, 'r'},
    {"bits", 1, 0, 'b'},
    {"parity", 1, 0, 'p'},
    {"stopbits", 1, 0, 's'},
    {"target", 1, 0, 't'},
    {"verbose", 0, 0, 'v'},
    {"help", 0, 0, 'h'},
    {"motor-max-speed", 1, 0, 'S'},
    {"max-frequency", 1, 0, 'F'},
    {"min-frequency", 1, 0, 'f'},
    {0,0,0,0}
};

static char *option_string = "d:r:b:p:s:t:vhS:F:f:";

static char *bitstrings[] = {"5", "6", "7", "8", NULL};

static char *paritystrings[] = {"even", "odd", "none", NULL};
static char paritychars[] = {'E', 'O', 'N'};

static char *ratestrings[] = {"1200", "2400", "4800", "9600", "19200", "38400", NULL};
static char *stopstrings[] = {"1", "2", NULL};


static void quit(int sig) {
    done = 1;
}


int match_string(char *string, char **matches) {
    int len, which, match;
    which=0;
    match=-1;
    if ((matches==NULL) || (string==NULL)) return -1;
    len = strlen(string);
    while (matches[which] != NULL) {
        if ((!strncmp(string, matches[which], len)) && (len <= strlen(matches[which]))) {
            if (match>=0) return -1;        // multiple matches
            match=which;
        }
        ++which;
    }
    return match;
}


void usage(int argc, char **argv) {
    printf("Usage:  %s [ARGUMENTS]\n", argv[0]);
    printf(
        "\n"
        "This program interfaces the Huanyang GT-series VFD to the LinuxCNC HAL.\n"
        "\n"
        "Required arguments:\n"
        "    -S, --motor-max-speed SPEED\n"
        "        The motor's max speed in RPM.  This must match the motor speed\n"
        "        value configured in VFD register P2.03.\n"
        "    -F, --max-frequency F\n"
        "        This is the maximum output frequency of the VFD in Hz.  It must\n"
        "        correspond to the maximum output frequency configured in VFD\n"
        "        register P0.03.\n"
        "    -f, --min-frequency F\n"
        "        This is the minimum output frequency of the VFD in Hz.  It must\n"
        "        correspond to the minimum output frequency configured in VFD\n"
        "        register P0.05.\n"
        "\n"
        "Optional arguments:\n"
        "    -d, --device PATH (default /dev/ttyS0)\n"
        "        Set the name of the serial device to use.\n"
        "    -r, --rate RATE (default 38400)\n"
        "        Set baud rate to RATE.  It is an error if the rate is not one of\n"
        "        the following: 1200, 2400, 4800, 9600, 19200, 38400\n"
        "    -b, --bits BITS (default 8)\n"
        "        Set number of data bits to BITS, must be between 5 and 8 inclusive.\n"
        "    -p, --parity PARITY (default none)\n"
        "        Set serial parity to one of 'even', 'odd', or 'none'.\n"
        "    -s, --stopbits {1,2} (default 1)\n"
        "        Set serial stop bits to 1 or 2.\n"
        "    -t, --target TARGET (default 1)\n"
        "        Set Modbus target (slave) number.  This must match the device\n"
        "        number you set on the Huanyang GT-series VFD.\n"
        "    -v, --verbose\n"
        "        Turn on verbose mode.\n"
        "    -h, --help\n"
        "        Show this help."
    );
}


// The Huanyang GT-series manual says that there must be either 3.5 bits
// or 3.5 bytes of silence before each packet (it says different things
// in different places).
void hy_modbus_sleep(void) {
    float seconds_per_bit = 1.0 / (float)baud;
    useconds_t useconds_per_bit = seconds_per_bit * 1000 * 1000;
    // useconds_t delay = useconds_per_bit * 8 * 3.5;
    useconds_t delay = useconds_per_bit * 8 * 10;
    usleep(delay);
}


int set_motor_on_forward(modbus_t *mb) {
    int r;
    uint16_t addr = 0x1000;
    uint16_t val = 1;

    for (int retries = 0; retries <= NUM_MODBUS_RETRIES; retries ++) {
        hy_modbus_sleep();
        r = modbus_write_register(mb, addr, val);
        if (r == 1) {
            return 0;
        }
        fprintf(stderr, "%s: error writing %u to register 0x%04x: %s\n", __func__, val, addr, modbus_strerror(errno));
        *haldata->modbus_errors = *haldata->modbus_errors + 1;
    }
    return -1;
}


int set_motor_off(modbus_t *mb) {
    int r;
    uint16_t addr = 0x1000;
    uint16_t val = 5;

    for (int retries = 0; retries <= NUM_MODBUS_RETRIES; retries ++) {
        hy_modbus_sleep();
        r = modbus_write_register(mb, addr, val);
        if (r == 1) {
            return 0;
        }
        fprintf(stderr, "%s: error writing %u to register 0x%04x: %s\n", __func__, val, addr, modbus_strerror(errno));
        *haldata->modbus_errors = *haldata->modbus_errors + 1;
    }
    return -1;
}


int set_motor_frequency(modbus_t *mb, float freq) {
    int r;
    uint16_t addr = 0x2000;
    uint16_t val;  // 100 * the frequency percentage
    val = (freq / max_freq) * 10000;

    for (int retries = 0; retries <= NUM_MODBUS_RETRIES; retries ++) {
        hy_modbus_sleep();
        r = modbus_write_register(mb, addr, val);
        if (r == 1) {
            return 0;
        }
        fprintf(stderr, "%s: error writing %u to register 0x%04x: %s\n", __func__, val, addr, modbus_strerror(errno));
        *haldata->modbus_errors = *haldata->modbus_errors + 1;
    }
    return -1;
}


int read_modbus_register(modbus_t *mb, modbus_register_t *reg) {
    uint16_t data;
    int r;

    for (int retries = 0; retries <= NUM_MODBUS_RETRIES; retries ++) {
        hy_modbus_sleep();
        r = modbus_read_registers(mb, reg->address, 1, &data);
        if (r == 1) {
            *reg->hal_pin = data * reg->multiplier;
            return 0;
        }
        fprintf(stderr, "%s: error reading %s (register 0x%04x): %s\n", __func__, reg->name, reg->address, modbus_strerror(errno));
        *haldata->modbus_errors = *haldata->modbus_errors + 1;
    }
    return -1;
}


void read_modbus_registers(modbus_t *mb) {
    for (int i = 0; i < num_modbus_registers; i ++) {
        (void)read_modbus_register(mb, &modbus_register[i]);
    }
}


modbus_register_t *add_modbus_register(modbus_t *mb, int address, const char *pin_name, float multiplier) {
    int r;
    modbus_register_t *reg;

    reg = &modbus_register[num_modbus_registers];

    r = hal_pin_float_newf(HAL_OUT, &reg->hal_pin, hal_comp_id, "%s.%s", modname, pin_name);
    if (r != 0) {
        return NULL;
    }

    reg->address = address;
    reg->name = pin_name;
    reg->multiplier = multiplier;

    read_modbus_register(mb, reg);

    num_modbus_registers ++;

    return reg;
}


int main(int argc, char **argv) {
    char *device;
    int bits;
    char parity;
    int stopbits;
    int verbose;

    int retval = 0;
    modbus_t *mb;
    int slave;
    struct timespec period_timespec;
    char *endarg;
    int opt;
    int argindex, argvalue;

    modbus_register_t *speed_fb_reg;

    done = 0;

    // assume that nothing is specified on the command line
    device = "/dev/ttyS0";
    baud = 38400;
    bits = 8;
    parity = 'N';
    stopbits = 1;

    verbose = 0;

    /* slave / register info */
    slave = 1;

    // process command line options
    while ((opt = getopt_long(argc, argv, option_string, long_options, NULL)) != -1) {
        switch (opt) {
            case 'd':   // device name, default /dev/ttyS0
                // could check the device name here, but we'll leave it to the library open
                if (strlen(optarg) > FILENAME_MAX) {
                    printf("ERROR: device node name is too long: %s\n", optarg);
                    retval = -1;
                    goto out_noclose;
                }
                device = strdup(optarg);
                break;

            case 'r':   // Baud rate, 38400 default
                argindex=match_string(optarg, ratestrings);
                if (argindex<0) {
                    printf("ERROR: invalid baud rate: %s\n", optarg);
                    retval = -1;
                    goto out_noclose;
                }
                baud = atoi(ratestrings[argindex]);
                break;

            case 'b':   // serial data bits, probably should be 8 (and defaults to 8)
                argindex=match_string(optarg, bitstrings);
                if (argindex<0) {
                    printf("ERROR: invalid number of bits: %s\n", optarg);
                    retval = -1;
                    goto out_noclose;
                }
                bits = atoi(bitstrings[argindex]);
                break;

            case 'p':   // parity, should be a string like "even", "odd", or "none"
                argindex=match_string(optarg, paritystrings);
                if (argindex<0) {
                    printf("ERROR: invalid parity: %s\n", optarg);
                    retval = -1;
                    goto out_noclose;
                }
                parity = paritychars[argindex];
                break;

            case 's':   // stop bits, defaults to 1
                argindex=match_string(optarg, stopstrings);
                if (argindex<0) {
                    printf("ERROR: invalid number of stop bits: %s\n", optarg);
                    retval = -1;
                    goto out_noclose;
                }
                stopbits = atoi(stopstrings[argindex]);
                break;

            case 't':   // target number (MODBUS ID), default 1
                argvalue = strtol(optarg, &endarg, 10);
                if ((*endarg != '\0') || (argvalue < 1) || (argvalue > 254)) {
                    printf("ERROR: invalid slave number: %s\n", optarg);
                    retval = -1;
                    goto out_noclose;
                }
                slave = argvalue;
                break;

            case 'v':
                verbose = 1;
                break;

            case 'h':
                usage(argc, argv);
                exit(0);
                break;

            case 'S':
                motor_max_speed = strtof(optarg, &endarg);
                if ((*endarg != '\0') || (motor_max_speed == 0.0)) {
                    printf("%s: ERROR: invalid motor max speed: %s\n", modname, optarg);
                    exit(1);
                }
                break;

            case 'F':
                max_freq = strtof(optarg, &endarg);
                if ((*endarg != '\0') || (max_freq == 0.0)) {
                    printf("%s: ERROR: invalid max freq: %s\n", modname, optarg);
                    exit(1);
                }
                break;

            case 'f':
                min_freq = strtof(optarg, &endarg);
                if ((*endarg != '\0') || (min_freq == 0.0)) {
                    printf("%s: ERROR: invalid min freq: %s\n", modname, optarg);
                    exit(1);
                }
                break;

            default:
                usage(argc, argv);
                exit(1);
                break;
        }
    }

    if (motor_max_speed == 0.0) {
        printf("%s: must specify --motor-max-speed\n", modname);
        exit(1);
    }

    if (max_freq == 0.0) {
        printf("%s: must specify --max-frequency\n", modname);
        exit(1);
    }

    if (min_freq == 0.0) {
        printf("%s: must specify --min-frequency\n", modname);
        exit(1);
    }

    if (min_freq > max_freq) {
        printf("%s: min frequency (%f) must be less than max frequency (%f)\n", modname, min_freq, max_freq);
        exit(1);
    }

    {
        struct sigaction sa;
        int r;

        sa.sa_handler = quit;
        sigemptyset(&sa.sa_mask);

        // libmodbus sometimes segfaults if its system calls get
        // interrupted by signals.  This works around it by automatically
        // restarting the system call instead of having it return EINTR.
        sa.sa_flags = SA_RESTART;

        r = sigaction(SIGINT, &sa, NULL);
        if (r != 0) {
            printf("failed to set SIGINT handler: %s\n", strerror(errno));
            exit(1);
        }
        r = sigaction(SIGTERM, &sa, NULL);
        if (r != 0) {
            printf("failed to set SIGTERM handler: %s\n", strerror(errno));
            exit(1);
        }
    }

    printf("%s: device='%s', baud=%d, parity='%c', bits=%d, stopbits=%d, address=%d\n",
           modname, device, baud, parity, bits, stopbits, slave);

    mb = modbus_new_rtu(device, baud, parity, bits, stopbits);
    if (mb == NULL) {
        printf("%s: ERROR: couldn't open modbus serial device: %s\n", modname, modbus_strerror(errno));
        goto out_noclose;
    }

    {
        struct timeval t;

        // Set the response timeout.
        t.tv_sec = 0;
        t.tv_usec = 16 * 1000;
#if (LIBMODBUS_VERSION_CHECK(3, 1, 2))
        modbus_set_response_timeout(mb, t.tv_sec, t.tv_usec);
#else
        modbus_set_response_timeout(mb, &t);
#endif

        // Disable the byte timeout so it just waits for the complete
        // response timeout instead.
#if (LIBMODBUS_VERSION_CHECK(3, 1, 2))
        t.tv_sec = 0;
        t.tv_usec = 0;
        modbus_set_byte_timeout(mb, t.tv_sec, t.tv_usec);
#else
        t.tv_sec = -1;
        modbus_set_byte_timeout(mb, &t);
#endif
    }

    retval = modbus_connect(mb);
    if (retval != 0) {
        printf("%s: ERROR: couldn't open serial device: %s\n", modname, modbus_strerror(errno));
        goto out_noclose;
    }

    modbus_set_debug(mb, verbose);

    modbus_set_slave(mb, slave);

    /* create HAL component */
    hal_comp_id = hal_init(modname);
    if (hal_comp_id < 0) {
        printf("%s: ERROR: hal_init failed\n", modname);
        retval = hal_comp_id;
        goto out_close;
    }

    haldata = (haldata_t *)hal_malloc(sizeof(haldata_t));
    if (haldata == NULL) {
        printf("%s: ERROR: unable to allocate shared memory\n", modname);
        retval = -1;
        goto out_closeHAL;
    }

    retval = hal_pin_float_newf(HAL_IN, &(haldata->period), hal_comp_id, "%s.period-seconds", modname);
    if (retval != 0) goto out_closeHAL;

    retval = hal_pin_float_newf(HAL_IN, &(haldata->speed_cmd), hal_comp_id, "%s.speed-cmd", modname);
    if (retval != 0) goto out_closeHAL;

    retval = hal_pin_float_newf(HAL_OUT, &(haldata->freq_cmd), hal_comp_id, "%s.freq-cmd", modname);
    if (retval != 0) goto out_closeHAL;

    retval = hal_pin_bit_newf(HAL_OUT, &(haldata->at_speed), hal_comp_id, "%s.at-speed", modname);
    if (retval != 0) goto out_closeHAL;

    retval = hal_pin_bit_newf(HAL_IN, &(haldata->spindle_on), hal_comp_id, "%s.spindle-on", modname);
    if (retval != 0) goto out_closeHAL;

    retval = hal_pin_u32_newf(HAL_OUT, &(haldata->modbus_errors), hal_comp_id, "%s.modbus-errors", modname);
    if (retval != 0) goto out_closeHAL;

    *haldata->period = 0.1;

    *haldata->freq_cmd = 0.0;
    *haldata->at_speed = 0;

    *haldata->modbus_errors = 0;

    modbus_register = (modbus_register_t *)hal_malloc(20 * sizeof(modbus_register_t));
    if (modbus_register == NULL) {
        printf("%s: ERROR: unable to allocate memory\n", modname);
        retval = -1;
        goto out_closeHAL;
    }

    if (add_modbus_register(mb, 0x3000, "freq-fb", 0.01) == NULL) {
        goto out_closeHAL;
    }

    if (add_modbus_register(mb, 0x3002, "dc-bus-voltage", 0.1) == NULL) {
        goto out_closeHAL;
    }

    if (add_modbus_register(mb, 0x3003, "output-voltage", 1.0) == NULL) {
        goto out_closeHAL;
    }

    if (add_modbus_register(mb, 0x3004, "output-current", 1.0) == NULL) {
        goto out_closeHAL;
    }

    speed_fb_reg = add_modbus_register(mb, 0x3005, "speed-fb", 1.0);
    if (speed_fb_reg == NULL) {
        goto out_closeHAL;
    }

    if (add_modbus_register(mb, 0x3006, "output-power", 1.0) == NULL) {
        goto out_closeHAL;
    }

    if (add_modbus_register(mb, 0x300a, "input-terminal", 1.0) == NULL) {
        goto out_closeHAL;
    }

    if (add_modbus_register(mb, 0x300b, "output-terminal", 1.0) == NULL) {
        goto out_closeHAL;
    }

    if (add_modbus_register(mb, 0x300c, "AI1", 1.0) == NULL) {
        goto out_closeHAL;
    }

    if (add_modbus_register(mb, 0x300d, "AI2", 1.0) == NULL) {
        goto out_closeHAL;
    }

    if (add_modbus_register(mb, 0x3010, "HDI-frequency", 1.0) == NULL) {
        goto out_closeHAL;
    }

    if (add_modbus_register(mb, 0x3014, "external-counter", 1.0) == NULL) {
        goto out_closeHAL;
    }

    if (add_modbus_register(mb, 0x5000, "fault-info", 1.0) == NULL) {
        goto out_closeHAL;
    }

    // Activate HAL component
    hal_ready(hal_comp_id);

    while (done == 0) {
        if (*haldata->period < 0.001) *haldata->period = 0.001;
        if (*haldata->period > 2.0) *haldata->period = 2.0;
        period_timespec.tv_sec = (time_t)(*haldata->period);
        period_timespec.tv_nsec = (long)((*haldata->period - period_timespec.tv_sec) * 1000000000l);
        nanosleep(&period_timespec, NULL);

        read_modbus_registers(mb);

        if (*haldata->spindle_on) {
            set_motor_on_forward(mb);
            *haldata->freq_cmd = (*haldata->speed_cmd / motor_max_speed) * max_freq;
            set_motor_frequency(mb, *haldata->freq_cmd);

            if ((fabs(*haldata->speed_cmd - *speed_fb_reg->hal_pin) / *haldata->speed_cmd) < 0.02) {
                *haldata->at_speed = 1;
            } else {
                *haldata->at_speed = 0;
            }
        } else {
            set_motor_off(mb);
            *haldata->at_speed = 0;
            *haldata->freq_cmd = 0.0;
        }

    }

    usleep(10 * 1000);
    set_motor_off(mb);

    retval = 0;	/* if we get here, then everything is fine, so just clean up and exit */
out_closeHAL:
    hal_exit(hal_comp_id);
out_close:
    modbus_close(mb);
    modbus_free(mb);
out_noclose:
    return retval;
}
